baseservice/iodevices/input_devices.py

```diff
@@ -0,0 +1,399 @@
+
+
+class InputDevice(metaclass=ABCMeta):
+    """
+    this is the base class for input devices
+    """
+    INPUT_DEVICE_NAME_HEADER = "__INPUT_DEVICE_NAME__"
+
+    def __init__(self, manager: 'InputDeviceManager', name: str):
+        """
+
+        :param manager: the input device manager that created this device
+        :param name: the name of this device
+        """
+        self._manager = manager
+        self._name = name
+
+    @property
+    def name(self) -> str:
+        """
+        :return: the name of this device
+        """
+        return self._name
+
+    @property
+    def manager(self) -> 'InputDeviceManager':
+        """
+        :return: the input device manager that created this device
+        """
+        return self._manager
+
+    def read_stream(self,
+                    timeout: Optional[float] = 0,
+                    with_transaction: bool = True) -> ReadStreamResult:
+        """
+        this method returns a message from the device. and makes sure that the input device name header is present
+
+        :param timeout: an optional timeout (in seconds) to wait for the device to return a message.
+        after 'timeout' seconds, if the device doesn't have a message to return, it will return (None, None)
+        :param with_transaction: 'True' if the device should read message within transaction,
+        or 'False' if the message is automatically committed
+
+        :return: a tuple of (Message, DeviceHeaders, Transaction) or (None, None, None) if no message was available.
+        the device headers, can contain extra information about the device that returned the message
+        """
+        message, device_headers, transaction = self._read_stream(timeout=timeout,
+                                                                 with_transaction=with_transaction)
+        if message is not None:
+            device_headers = device_headers or {}
+            device_headers.setdefault(self.INPUT_DEVICE_NAME_HEADER, self.name)
+
+        return message, device_headers, transaction
+
+    @abstractmethod
+    def _read_stream(self,
+                     timeout: Optional[float] = 0,
+                     with_transaction: bool = True) -> ReadStreamResult:
+        """
+        this method returns a message from the device (should be implemented by child classes)
+
+        :param timeout: an optional timeout (in seconds) to wait for the device to return a message.
+        after 'timeout' seconds, if the device doesn't have a message to return, it will return (None, None)
+        :param with_transaction: 'True' if the device should read message within transaction,
+        or 'False' if the message is automatically committed
+
+        :return: a tuple of (Message, DeviceHeaders, Transaction) or (None, None, None) if no message was available.
+        the device headers, can contain extra information about the device that returned the message
+        """
+        pass
```

Comment:
Its fine for optional functions to have this default implementation. However, the rest should raise a NotImplementedError

Reply:
i've seen in other places, that @abstractmethod methods, just define 'pass'. is there a convention saying otherwise?